### ISSUE TITLE
--state-values-set does not work properly in gotmpl value files

### ISSUE THREAD

**Issue by vladimir-avinkin** [helmfile/helmfile#700]
### Operating system

MacOS 12.1

### Helmfile Version

0.150.0

### Helm Version

v3.9.4

### Bug description

See the linked github repo's README file

### Example helmfile.yaml

See the linked github repo's README file.

### Error message you've seen (if any)

None

### Steps to reproduce

https://github.com/vladimir-avinkin/helmfile-bug-repro

### Working Helmfile Version

Haven't found one

**Comment by yxxhero** [comment]
@vladimir-avinkin quick solution:
```
environments:
  default:
    values:
      - base.yaml
      - override.yaml
# key change
--- 
repositories:
  - name: bitnami
    url: https://charts.bitnami.com/bitnami

releases:
  - name: test
    chart: bitnami/nginx
    namespace: {{ .Values.ns }}
    values:
      - values.yaml.gotmpl
```

**Comment by yxxhero** [comment]
@vladimir-avinkin fixed: https://github.com/helmfile/helmfile/pull/705

**Comment by mumoshu** [comment]
Hey @vladimir-avinkin and  @yxxhero! This is actually an expected behavior. Helmfile's `--state-values-set` sets the values used to render/merge/compose the environment values. In other words, values provided via `--state-values-set` are both inputs and default values for the rendered environment values. It's going to be a breaking change which we don't want to make.

That said, do you still want a new feature to accommodate your use case? What if we had a new flag `--state-values-override` that works like you desired? Shall we name it just `--set`? Also note that Helmfile's `--state-value-set` is conceptually different from Helm's `--set`- Helm does not have dynamic values while Helmfile does have it via the environment values templating. The former has no notion of "inputs" while the latter does have it.

**Comment by mumoshu** [comment]
TL;DR; you guys seem to have expected values set via --state-values-set to be “overrides”. Actually, they are more like “defaults” and “inputs”. I don't like making breaking changes. However, I'm super open to adding a new feature instead to address your use case.

**Comment by yxxhero** [comment]
Oh. Got it.

**Comment by yxxhero** [comment]
@mumoshu but if we add `---` between `environments` and `releases`, it has different behavior.

**Comment by mumoshu** [comment]
@yxxhero As far as I remember, the vaules provided via `--state-values-set` should be treated as the defaults and the inputs for the first part of your helmfile.yaml when each part is separated by `---`. Have you seen something different than that?

**Comment by yxxhero** [comment]
@mumoshu from your explanation. I think the behavior is expected. So we should add more help messages for `--state-values-set`.

**Comment by vladimir-avinkin** [comment]
> Hey @vladimir-avinkin and @yxxhero! This is actually an expected behavior. Helmfile's `--state-values-set` sets the values used to render/merge/compose the environment values. In other words, values provided via `--state-values-set` are both inputs and default values for the rendered environment values. It's going to be a breaking change which we don't want to make.
> 
> That said, do you still want a new feature to accommodate your use case? What if we had a new flag `--state-values-override` that works like you desired? Shall we name it just `--set`? Also note that Helmfile's `--state-value-set` is conceptually different from Helm's `--set`- Helm does not have dynamic values while Helmfile does have it via the environment values templating. The former has no notion of "inputs" while the latter does have it.

To be honest I'm quite confused here. I was expecting `--state-values-set` to be a CLI version of environments.*.values the same way `--set` is the CLI version of the regular helm values files. And it's a convention in most unix/cloud tools that CLI flags have higher priority, as it is in helm itself for example.

Even if it's not the case, I don't get why would the same environmental value (that's what we access when we use .Values after all) be different in the context of helmfile.yaml, and gotmpl templates. Wouldn't you expect to see test2 in both contexts in my example?

Nevertheless, for my specific use case, I just need a way to pass `environments.*.values` through cli somehow. It's because my helmfile has defaults specified in files, but I want to be able to pass dynamically generated values in the CI pipeline, to pass for example, a custom temporary domain name, so I guess your proposal would work for me.

**Comment by yxxhero** [comment]
@mumoshu  see origin PR. https://github.com/roboll/helmfile/pull/644. it should be as override.  WDYT?

**Comment by mumoshu** [comment]
@vladimir-avinkin I understand your confusion. I would have been confused the same way if I came to Helmfile for my first time.

But I vaguely remember folks requested how to provide inputs to dynamically generate environment values and currently `--state-values-set` is the only way for providing inputs. If it was "overrides", there's no way to provide the inputs.

> Nevertheless, for my specific use case, I just need a way to pass environments.*.values through cli somehow. It's because my helmfile has defaults specified in files, but I want to be able to pass dynamically generated values in the CI pipeline, to pass for example, a custom temporary domain name, so I guess your proposal would work for me.

I take this as 100% valid use case. What if we added another flag like `--set` which might look more aligned with what helm provides?

**Comment by mumoshu** [comment]
> Even if it's not the case, I don't get why would the same environmental value (that's what we access when we use .Values after all) be different in the context of helmfile.yaml, and gotmpl templates. Wouldn't you expect to see test2 in both contexts in my example?

@vladimir-avinkin Good point! You should have access to the same environment values in both helmfile.yaml and release values.yaml.gotmpl files. If there's any divergence, that might be a bug 🤔 Let me see what's going on here.

**Comment by mumoshu** [comment]
@vladimir-avinkin In the meantime, would you mind separating the environments and the releases parts with `---`? You might be unintentionally using our legacy "double-rendering" feature which we're trying to deprecate in the upcoming Helmfile V1. `---` is the key to make double-rendering virtually "no-op". https://github.com/helmfile/helmfile/blob/main/docs/proposals/towards-1.0.md#forbid-the-use-of-environments-and-releases-within-a-single-helmfileyamlgotmpl-part

**Comment by vladimir-avinkin** [comment]
Yep, it seems like separating with `---` resulted in an expected (for me) behavior. 👍 

If you want to introduce another cli option it is also fine with me, albeit a bit weird :) But I will leave it to you
